RFC: Simpler array hashing

mbauman opened the pull request:
This is a straw-man implementation of a simpler array hashing scheme.  It's very basic and has room for further optimizations, but it's intention is to provide a starting point as an alternative to JuliaLang/julia#25822.  In short: This hashes the size of the array and then the first three distinct elements and their linear indices and then the last three distinct elements and their linear distance from the end of the array.

The exact scheme here is open to bikeshedding -- we could use more or less distinct elements.  I use "distinct elements" as a way of ensuring that all relatively empty sparse arrays don't hash similarly, and I hash elements at both the beginning and end of the array because they're the simplest to get to and final elements will be the "most expensive" to discover that they differ if we have to fall back to an equality check. The most complicated part is keeping track of what you hashed in order to prevent running through the entire array twice (once forwards and once backwards).

nalimilan commented:
Thanks for this. I certainly appreciate the simplicity of this approach compared with the current one. I think we should keep in mind a third possible approach, though, which is to just drop O(1) hashing for ranges (it's not clear that's very useful in practice) and use a simple solution similar to what was implemented before https://github.com/JuliaLang/julia/pull/16401. As a hybrid solution, we could also continue to hash all elements for multidimensional arrays (as done both before and after https://github.com/JuliaLang/julia/pull/16401), and only change how vectors are hashed, since that's what is needed to hash ranges in O(1) time.

Assuming we adopt the general approach from this PR (i.e. do not hash all elements nor all differences between subsequent elements, but only a part of the information), several other quantities could be included in the hash, with the objective of making collisions less likely when modifying a single entry in the middle of the array:
- hash the sum of all elements
- hash the number of nonzero elements
- hash the indices of series of nonzero elements (i.e. ranges of indices rather than individual indices)

These quantities have the advantage that they can be computed in O(1) time for (integer, at least) ranges and relatively efficiently for sparse matrices. But they require going over the full array, making the computation of hashes O(N) for general arrays while your PR is currently O(1).

Overall I think we should identify a few typical use cases to decide which approach is the best one. Without that it difficult to know what's the best tradeoff between computation cost and probability of collision. In general I'm a bit concerned about the fact that changing the value of entries in the middle of the array wouldn't change the hash: ideally collisions shouldn't happen in such trivial cases IMHO. But maybe that's fine in practice and the reduction in hashing cost is worth it. I'll just leave a pointer to https://github.com/JuliaLang/julia/issues/25220, which AFAICT isn't a very correct use of array hashing, but which is going to become even more problematic if we make collisions much more likely (it could use a custom algorithm if needed).

Finally, let me note that we still have the possibility of continuing to hash all elements for multidimensional arrays (as was done before and after https://github.com/JuliaLang/julia/pull/16401), and only change how vectors are hashed, since that's what is needed to hash ranges in O(1) time.

Doing some research, I've found only two examples of array hashing in main languages:
- C# only hashes the last 8 elements in an array by default: https://stackoverflow.com/a/37123103/2413179
- Java hashes all elements: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/Arrays.java#Arrays.hashCode%28short%5B%5D%29

mbauman commented on base/abstractarray.jl:
```diff
@@ -1945,111 +1945,34 @@ pushfirst!(A, a, b, c...) = pushfirst!(pushfirst!(A, c...), a, b)
-# Efficient O(1) method equivalent to the O(N) AbstractArray fallback,
-# which works only for ranges with regular step (RangeStepRegular)
-function hash_range(r::AbstractRange, h::UInt)
-    h += hashaa_seed
```
Note to self: I forgot to include the `hashaa_seed` in the computation

JeffBezanson commented on base/abstractarray.jl:
```diff
@@ -1945,111 +1945,34 @@ pushfirst!(A, a, b, c...) = pushfirst!(pushfirst!(A, c...), a, b)
+# But we cannot define those here, due to bootstrapping. They're defined later in iterators.jl.
+function hashndistinct(A, I, n, hx)
+    _hashcheckbounds(A, I)
+    seen = Set()
```
Would probably be better to use a Vector for this since the number of items is small, and it will avoid hashing them multiple times.

mbauman replied:
Yeah, I was thinking of using a specialized 3-vector to avoid allocations entirely, but figured we  could decide on the direction and exact behavior first before optimizing it further.

JeffreySarnoff commented:
This is a good optimization.  These are some hash admixes worth consideration.

(0) Hashing the hashed name of the variable that references or holds the collection.

then in some manner, one from one or more of these pairs may help

(a) the collection type
(α) the eltype

(b) the shape (dims) # reducingly hash each dim size
(β) the length (prod(dims))

(k) the initial 1st, 2nd, 7th and final 1st (end), 2nd, 7th (end-6)  (or missings)
(κ) the initial and final two and initial and final fn(length)th element
      `fn(length) = mod(maxmin(length, 61)...,) + 1`

nalimilan commented:
@JeffreySarnoff (0), (a) and (α) do not matter for `isequal`, so they definitely cannot enter the hash.

mbauman commented:
So the options and their tradeoffs here are:

* If we don't hash every single element, it's easy to mutate an array and have its hash remain the same.  That's just fine, though, since hash collisions are allowed and will fall back to equality checks.  We just don't want it to happen too often, since those equality checks could be expensive to repeatedly perform.
* If we hash every single element, it's very expensive for sparse matrices to compute every single element. Making hashing ranges O(N) might make an otherwise cromulent structure like `1:typemax(Int)` stall till the heat death of the universe if you happen to put it in a set.
* If we do something clever, we have to be doubly clever in order to deal with heterogeneous arrays of things that might not support the cleverness or might overflow/underflow/change precision (e.g., subtraction for diff, addition for sum, promotion and precision difficulties in both cases).  The only sort of cleverness I'd advocate for is the kind that only relies upon hashing and equality — like run-length encoding or distinct elements.

I'm becoming more convinced that an O(1) hashing method is the way to go here.  I'd happily extend the number of elements to help convince you.

nalimilan commented:
I generally agree, the current approach is too brittle and complex. I still a bit sad to move to an approach which makes it so easy to get collisions. Maybe I just need some time to accept this. ;-)

Another less fragile possibility which occurred to me would be to use another property of ranges, which is less strict than computing differences between subsequent elements but also less problematic to check: the fact that ranges are monotonically increasing or decreasing. We could continue to hash all elements for vectors which are not sorted and multidimensional arrays (keeping RLE for efficiency with sparse matrices of course), but use the O(1) approach for sorted vectors (for element types which support `isless`). In practice, it should be quite rare that a user would hash several ordered vectors of the same length and which start and end with the same elements, so this should dramatically reduce the risk of collisions.

JeffreySarnoff commented:
I agree that loosing discrimination is a negative (there are algorithms that thrive on some probable assurance of differents hashing differently, so things need be visited once to be useful with other things of a group). 

Increasing the span at front and back that are hashed, and speckle-ing some of the ith quartile positions (or a few thereabout) should do a decent of job of discrimination, especially where the length>>some mod a_length_relevant_prime is used to pick the span between successive hash-sampled values (or value pairs).

Or for speed, keep a sequence of indices to pluck and let them wrap on the length.

JeffBezanson commented:
One option might be to use exponentially-spaced indices, and hash `log(n)` points.

JeffreySarnoff commented:
... and interweave those indicies, taking them from [1] increasing and from [end] decreasing,  to disambiguate small vectors and similar arrays more successfully?

JeffreySarnoff commented:
```
struct EntityHash{T<:Union{UInt64, Missing}}
    eachentity::UInt64  # every hashable thing has this valued
    collective::T       # an aggregative thing may have this valued
                        #    lazily done only if needed or where
                        #    the e.g. write stream does the calc
end

Things that are items or elements have sizeof(EntityHash) == sizeof(UInt64)
Things that are collectives and are more fully hashed have twice that size
```

martinholters commented:
>We could continue to hash all elements for vectors which are not sorted and multidimensional arrays (keeping RLE for efficiency with sparse matrices of course), but use the O(1) approach for sorted vectors (for element types which support `isless`).

Fails for the `totally_not_five` from #26034. (Not that that example seems particularly relevant, but the fact the it is so easy define an object which breaks the approach is concerning.)

mbauman commented:
Taking `log(n)` hashes is quite compelling — particularly if we increase distance between hashes as we iterate from the back of the array, since that weights the elements involved in the hash towards the "most expensive" to differentiate with isequal.  The major downside I see there is that many sparse matrices will have fewer than log(n) stored values.

What if we used a combination of these approaches?

* Given array `A`, consider the list of hashes `hxs = [hash(A[end]); [hash(A[i]) for i in lastindex(A)-1:firstindex(A) if hash(A[i]) != hash(A[i+1]) && A[i] != A[i+1]]`
* `hash(A)` is `sum(hxs[[2^i for i=0:floor(Int, log2(end))]])`

This has the slightly strange effect that we compute the hash of every element for `Array` but don't use them all in the computation, but I think it's way better than taking the difference of every element.  The advantages are: fast iteration through sparse arrays (often just alternating between a hash of zero and a hash of a stored value), and we only need to hash `O(N)` elements in computed arrays where we know successive elements are not equal.

mbauman commented:
The proposal in my last post still leaves a bad taste in my mouth.  If we're going to do `log(N)`, let's make it work for all array types.  Here's another idea:

* Iterate the array backwards
* Add the hashes the first 3 (or so) distinct values you see, stopping at index `j`
* Launch into a blind `log(N)` hashing regime starting at index `j`, adding `sum(hash.(A[[j - 2^i for i in 0:floor(log2(end-j)))]))`

StefanKarpinski commented:
I'm a bit concerned that hashing `log(N)` entries in a sufficiently sparse array will tend to catch none of the non-zeros.

mbauman commented:
Right — that's my second proposal.  In short:  Change this PR (which does O(1) hashing of some small number of distinct elements from the front and back) to ignore the forward iteration and augment the backwards iteration with log(N) hashes after it's hashed the small number of distinct elements.

nalimilan commented:
That's an interesting solution. There are a few things I don't understand:
- Why do you need to iterate backwards?
- Why do you compare the hashes of subsequent elements before comparing them directly?
- Why sum the hashes rather than combine them with `hash`?

mbauman commented:
Great questions:

* I iterate backwards because of the asymmetry between `==` and `hash`.  In the case of a hash collision, we fall back to `==`.  By default and most commonly, `==` iteratively compares its elements walking forwards through the array.  This means that — if the arrays are indeed different — it'll probably find that faster.  I would bet this is why C# hashes the last 8 elements.
* I compare hashes because I had been thinking we were already computing them — either as part of the total array hash or in a Set to hold the distinct elements.  Might as well use them!
* I sum `h += hash(A[i])` instead of chaining with `hash(A[i], h)` in order to allow for the reuse of the the hashes.

But you're right — it's probably not worth hashing elements that don't get used in the total hash, and we probably don't want to use a `Set` for such a limited number of elements and comparisons.

nalimilan commented:
I think we should proceed with any reasonable solution so that the API is stabilized and `BigFloat` can be moved to stdlib (https://github.com/JuliaLang/julia/issues/25716). We can always switch to a different approach in the future since it's not breaking.

JeffreySarnoff commented:
> I'm a bit concerned that hashing log(N) entries in a sufficiently sparse array will tend to catch none of the non-zeros.

Unclear that log(N) is the "hatrack" .. the same would be true of sqrt(N) samples.

Milan's push to get this one done so other contingencies can move along makes sense.  It all sounds non-blocking and pathway providing. 

_Maybe it is worthwhile to keep a short vector of linearized indices to nonzero values to provide informed indirection.  32 or 64 Int64s could do that for a sparse array of 10 trillion elements.  If the data structure knows the index of its first and its final nonzero entries, two of those indices could be rewritten when that internal state is rewritten.  Two can get two more (the next innermost) and the structure may find that useful too, giving the hash at worst 4 nonzero values and their linearized indexings -- 8 nonzero Ints to hash or better_

mbauman commented:
Yes, let's resurrect this.  I have a WIP branch locally where I started implementing https://github.com/JuliaLang/julia/pull/26022#issuecomment-366308770 a while ago, but I stalled on it (and then forgot about it) because I didn't like how hard it was to either implement or describe.  I'll take a second look this afternoon, and see if I can find some simplifications.

StefanKarpinski commented:
Here's another thought: hash arrays in such a manner that if an array approximates an range, it hashes like it and just take the hit of those two objects being in the same bucket. After all, what are the chances that you're hashing both arrays and ranges together and have an array that is almost equal to a range but not quite exactly equal to it? That's the only bad case with such an approach.

oscardssmith commented:
won't that fail because any bucket will have an edge that will have the same precision problems that your were trying to avoid?

JeffreySarnoff commented:
XORing the reinterpret(UInt64s, array_element_i) to summarize the ith n-dimensional element in an array is cheap and reasonable, again moving the even dimensional constituents (e.g. rotating the reinterpreted float bit pattern) and perhaps the odd ones the other way (12 bit rotations would do for Float64s) as they are hash accumulated (xored in) makes better oatmeal.  This should be done with the intitial four or, better, seven, or better 17  arrayed entries (for arrays of enough constituents).    Another 7 or (other than 7) should come from the end of a generic uprocessor's contiguous cache line -- if there are enough (faster get, still farish from start of array).  Whatever you want to do to glom onto more of the contents should be done pairwise with respect to a common cache line size (which may be Level1 cache for machine A and Level3 cache for machine B).  Unless you are going to replicate the final few entries, or their hashed agglomeration, at the head of the array structure, requireing that the hashing obtain the last 17 entries will impose unreasonable thrashing for big and partially external data.

nalimilan commented:
> Here's another thought: hash arrays in such a manner that if an array approximates an range, it hashes like it and just take the hit of those two objects being in the same bucket. After all, what are the chances that you're hashing both arrays and ranges together and have an array that is almost equal to a range but not quite exactly equal to it? That's the only bad case with such an approach.

I don't think the problem is really approximate vs. exact equality. Currently we hash floating point ranges in O(N) time, which eliminates the problem of approximations (assuming that we don't really care about the time it takes to hash a floating point range). But we still have issues due to the use of `-`: #27865, #26011, https://github.com/JuliaLang/julia/pull/16401#issuecomment-355210342. These are really the most annoying problems with the current approach. Restricting them to `Number` would certainly mitigate (or even fix?) the problem (see https://github.com/JuliaLang/julia/pull/25822). Using approximate equality would be interesting, but I rather see this as a refinement which would allow hashing floating point ranges in O(1) time, which can be done in 1.x without changing the overall approach.

OTOH, the approach from this PR has the advantage of simplicity, with the main drawback being the increased risk of collisions. If we really want to give a last chance to the current design, we could merge something like https://github.com/JuliaLang/julia/pull/25822 for 0.7, and if issues keep being reported, switch to this PR for 1.0. Since it's much simpler than the current system, it shouldn't introduce any new problems.

JeffreySarnoff commented:
It is easy to understand doing this now.  And it is unclear what benefit accrues to giving the current way one more opportunity to persuade people to want to do this. It should be easy to adjust the less-collisions knob should there be desire to do that when there is experience using this way.

mbauman committed:
Simplify hashing of AbstractArray

mbauman commented:
Alright, I've pushed my latest iteration here.  See the inline comments for a detailed explanation.  In short:

* Hash axes
* Hash three distinct key-value pairs, working backwards
* Starting at the last distinct key, begin blindly skipping elements, using the Fibonacci sequence to generate the skips.

Things I like:

* We use `keys`, `findprev`, and `LinearIndices` in a manner that is fully generic but still supports efficient cartesian indexing.
* Custom array types will probably not need to implement `hash` — instead, they `findprev`.
* It's a very simple implementation

The one thing I don't particularly like is that I iterate backwards.  I spent some time trying to put together a reverse fibonacci sequence so we could work forwards (but still have a higher density towards the end), but it was complicated, had trouble bootstrapping, was more prone to issues with overflow and other custom integer types, and didn't as seamlessly incorporate the distinct elements.  I find the simplicity of working backwards to far outweigh the slight (and unmeasured) advantage to cache locality that (I think) working forwards would have.

chethega commented:
Just recalling the alternate approach [https://discourse.julialang.org/t/alternate-approach-to-o-log-n-integer-range-hashing/9065/7](https://discourse.julialang.org/t/alternate-approach-to-o-log-n-integer-range-hashing/9065/7). When doing that, all containers would hash kinda normally, except that integer ranges and sparse containers would have a shortcut (permitting non-zero default elements of sparse containers, and permitting an unsorted list of non-default elements). 

This would necessitate a split of the current hashing construct, though: Conceptually, we would need a `compress: (UInt, UInt)->UInt` that approximates a random function and is globally fixed (possibly keyed, in order to prevent hash-collision-DoS), and a `normalize: Any->UInt` that has few collisions, is well-defined modulo `isequal`, and is affine-linear for small integers (this allows the fast range hashing).

mbauman committed:
Fixup type instability due to boxes

mbauman commented:
Ok, I've fixed the type instability (#15276), and I'm pretty happy with this now.  I've done some cursory testing of the performance impacts here by [filling and emptying dictionaries](https://gist.github.com/mbauman/0da1320e403f518b5450a78ad1f2ff4a) of arrays — overall things look pretty good but note that this can cause dictionaries to change algorithmic complexity.  I contrived the last two cases such that they're hash collisions but unequal; this is a worst case and leads to `O(length(A)*length(dict)^2)` performance.  Of course, it's also possible to construct a set of arrays that will all have colliding hashes with the current system, too — it's just a bit harder to do.

keys | fill_before | empty_before | fill_after | empty_after
-----|----------|---------|----------|------
`[(1:10) .- i for i in 1:100000]` | 0.01 | 0.008 | 0.04 | 0.03
`[(1:2^60) .- i for i in 1:100000]` | 0.01 | 0.005 | 0.1 | 0.07
`[(1:10)./i for i in 1:100000]` | 0.08 | 0.04 | 0.05 | 0.03
`[(1:2^60)./i for i in 1:100000]` | ~Inf | ~Inf | 0.4 | 0.2
`[rand(1000, 100) for _ in 1:100]` | 0.1 | 0.07 | 0.00007 | 0.0001
`[rand(1000, 100) for _ in 1:100000]` | 100 | 70 | 0.3 | 0.2
`A = rand(1000,100); [setindex!(copy(A), i, length(A)-20) for i in 1:100]` | 0.1 | 0.07 | 1.1 | 0.0002
`A = rand(1000,100); [setindex!(copy(A), i, length(A)-20) for i in 1:1000]` | 2 | 0.7 | 100 | 0.007

vtjnash commented:
> O(length(A)*length(dict)^2) performance

I think this is usually dealt with by using a more resilient dictionary algorithm (this is the problem case for the linear probing scheme we currently use) or a more tailored data-structure for the problem (such as a Trie). I'm surprised it's quite that bad though – typically the goal is to have the worst case for hash collision be to turn the lookup into a linear-scan (without the square power). Where does the `^2` power derive from?

vtjnash commented on base/abstractarray.jl:
possibly should use `Ref{eltype(A)}(v1)` here, or just open-code the loop (to make heterogeneous arrays fast). the same would apply to `h = hash(i=>v1, h)` though (vs. writing it as `h = hash(i, hash(v1, h))`), so I'm not sure of the importance of it

mbauman replied:
I added a simple type-unstable case to the matrix above — this is already smoking the status quo (largely by simply touching fewer unstable elements). I get your point about constructing the `Pair`s, but I don't fully understand the `Ref` in the closure.  In any case, this should be a relatively small constant cost.

mbauman commented:
> Where does the ^2 power derive from?

This is the complexity for inserting the whole set of hash-equal items into the dictionary.  For each item, it must compare `==` against all existing items in the dictionary.

nalimilan commented:
Thanks! The code and the benchmarks look good to me. Of course it's somewhat slower for ranges, and much slower when there are collisions, as expected. But the main question which is hard to address is whether collisions are frequent in practice. If that'd really the case (which isn't very likely), an ideal approach would maybe to have `Dict` use two hashes: this one as a first default pass, and a systematic hashing of all elements as soon as a collision happens (ignored if ranges are involved). If collisions are rare, this PR should make things quite faster since much fewer elements are hashed.

BTW, if this solution is retained, we'll need to do something about https://github.com/JuliaLang/julia/issues/25220. Maybe just use the old algorithm there, to limit the risk that the array wouldn't be updated because of collisions (but of course it's not completely correct either).

mbauman commented on base/abstractarray.jl:
Sure, this could be `@inbounds` but I wouldn't expect it to make a difference in performance — profiling hashing a Float64 array shows `checkbounds` accounts for 8 samples out of ~5000.  I'm inclined to leave it.

> Also, have you considered hashing the first entry which differs from the previous/next one

Yes, something along those lines is compelling, but I had avoided it due to the challenge of efficiently skipping linear offsets across cartesian indices.  Now that I have coded that computation here, though, perhaps it wouldn't be so bad. I'll see if it can be done reasonably.

mbauman committed:
New design: interleave fibonacci skips with finding non-repeating elements

    Goal: Hash approximately log(N) entries with a higher density of hashed elements
    weighted towards the end and special consideration for repeated values. Colliding
    hashes will often subsequently be compared by equality -- and equality between arrays
    works elementwise forwards and is short-circuiting. This means that a collision
    between arrays that differ by elements at the beginning is cheaper than one where the
    difference is towards the end. Furthermore, blindly choosing log(N) entries from a
    sparse array will likely only choose the same element repeatedly (zero in this case).

    To achieve this, we work backwards, starting by hashing the last element of the
    array. After hashing each element, we skip the next `fibskip` elements, where
    `fibskip` is pulled from the Fibonacci sequence -- Fibonacci was chosen as a simple
    ~O(log(N)) algorithm that ensures we don't hit a common divisor of a dimension and
    only end up hashing one slice of the array (as might happen with powers of two).
    Finally, we find the next distinct value from the one we just hashed.

mbauman commented:
Thanks @nalimilan — I like the design you proposed in https://github.com/JuliaLang/julia/pull/26022#discussion_r201595259 much better.

mbauman commented on base/abstractarray.jl:
```diff
@@ -2067,40 +2067,46 @@ function hash(A::AbstractArray, h::UInt)
         fibskip, prevfibskip = fibskip + prevfibskip, fibskip
+
+        # Find a key index with a value distinct from `elt` -- might be `keyidx` itself
+        keyidx = findprev(!isequal(elt), A, keyidx)
```
Yes, that's exactly right.  I think it makes the behavior a little more robust — otherwise the hashes of sparse arrays with a nonzero last element will more likely hash the same.  I also think it's most likely that diagonals of sparse matrices are filled.

That said, this is now clearly not hashing enough elements.  The test failures are from four-element arrays colliding.  Gotta slow down the exponential a little bit.

nalimilan replied:
> Yes, that's exactly right. I think it makes the behavior a little more robust — otherwise the hashes of sparse arrays with a nonzero last element will more likely hash the same. I also think it's most likely that diagonals of sparse matrices are filled.

Sorry, I'm not sure I follow. Could you develop?

> That said, this is now clearly not hashing enough elements. The test failures are from four-element arrays colliding. Gotta slow down the exponential a little bit.

Agreed.

nalimilan replied:
BTW, why use `findprev`, which requires all this keys vs. linear indices dance, instead of a plain loop? I imagine one reason could be that `findprev` could have a specialized method for sparse arrays which would skip empty ranges, but currently that's not the case. Is that what you have in mind?

mbauman committed:
Slow down the Fibonacci sequence

mbauman commented:
Ok, I've slowed down that exponential — by quite a bit.  I now only grab the next number in the Fibonacci sequence after every 4096 hashes.  I chose such a large number by considering:

* The large anticipated penalty for a hash collision. I wanted to maximize the size of the array that gets 100% of its elements hashed… especially in light of #25220.
    * Every element in arrays smaller than length 4096 (64x64 or 8x8x8x8) is included in the hash computation.
    * For an array of size 8192 with no sequentially repeated elements, 75% of its elements are hashed: every other element is hashed in the first half of the array, and every element is hashed in the second half.
* The time it takes to hash ranges with `typemax(Int)` elements as they used to be instantaneous — so I wanted to keep this at least somewhat tractable.  With this choice on my beefy machine, it takes ~0.5 s to hash `(1:typemax(Int64))/pi`, and ~10 ms to hash `1:typemax(Int64)`.
    * For an array of a billion elements with no repeats, 99,915 elements are hashed (10^5/10^9 — 0.01%)
    * For an array of `typemax(Int)` elements with no repeats, 276,303 elements are hashed (10^5.5/10^19).

This is the balance point that seems reasonable to me — a power of two just because it's slightly cheaper to perform the mod.

nalimilan commented:
Makes sense. I wonder whether we couldn't find a function which increases slowly for common sizes, but quite faster when sizes get really large. Or maybe we could just apply an arbitrary threshold beyond which we stop hashing entries? A `Vector{UInt8}` with `typemax(Int)` elements takes 8e6 TB, which hardware won't be able to handle for decades or centuries. It wouldn't be terrible if we stopped at `typemax(Int)/1e6` elements.

Liozou commented on base/abstractarray.jl:
Since performance is important here, why not replacing the `mod` with explicit bit twiddling? In this case I would suggest replacing this line with:
```
if (Int==Int64 && n<<52==0) || n<<20==0
```
That's because 4096 = 2^12 and 52 = 64 - 12 and 20 = 32 - 12. Also the `Int==Int64` part will get inlined so this will be effectively replaced by either `n<<52==0` or `n<<20==0`.

Liozou replied:
I always forget about the `rem` and `%`

JeffBezanson commented:
Bump. Could we add the test cases from #27865, #26011, and https://github.com/JuliaLang/julia/pull/16401#issuecomment-355210342 and merge this?

mbauman committed:
Add tests

Fixes #27865 and fixes #26011.

mbauman committed:
Use rem instead of mod

mbauman committed:
Also add test for #26034

Fixes #26034

mbauman commented:
Ok, test cases pushed.  There are, of course, many other schemes that we could dream up here… but triage was in support of merging this as it stands.

To help avoid #25220, I'm putting together another PR that will basically make Distributed use its own hashing system — the constraints are a little different there since we don't need to worry about hash equality across different types.  I don't think the two need to be merged at the same time, but I should have that ready to go by tomorrow.

mbauman committed:
Fixup test case for 32-bit machines
